### PR TITLE
Issue27 refactor statNode and define property model

### LINKED ISSUE #7: [Feature] Data-source and data property listener
## Issue Description

Type: *feature request*

### Describe what feature you want

Data-source and data property listener (`SentinelProperty` and `PropertyListener` in Java). Some considerations:

- We might need to support multiple `SentinelProperty` (with different parsers) shared by the single data-source.

### LINKED ISSUE #27: [Enhancement] refactor base statNode
## Issue Description

refactor statNode

### PR DESCRIPTION
**Pull request by louyuting**
### Describe what this PR does / why we need it
enhance statNode.
Define ReadStat and WriteStat to decouple the node operation.

define basic property update model dynamically.

### Does this pull request fix one issue?

Fixes #27 and also resolves #7 

### Special notes for reviews

### CODE REVIEW

File: core/stat/resource_node.go [hunk 1]
```diff
@@ -45,3 +45,7 @@ func (n *ResourceNode) ResourceType() base.ResourceType {
 func (n *ResourceNode) ResourceName() string {
 	return n.resourceName
 }
+
+func (n *ResourceNode) RealBucketLeapArray() *sbase.BucketLeapArray {
```
Comment (sczyh30) [inline core/stat/resource_node.go]: Maybe this is not necessary? The inner LeapArray should be invisible to users.

Comment (louyuting) [reply core/stat/resource_node.go]: It is used in circuit breaker.
```go
resNode := stat.GetOrCreateResourceNode(rule.ResourceName())
bla := resNode.RealBucketLeapArray()
```
when we new SlidingWindowMetric, we need pass parameter "*BucketLeapArray".

Comment (louyuting) [reply core/stat/resource_node.go]: There may be other ways to implement this requirement.

Comment (sczyh30) [reply core/stat/resource_node.go]: It's okay for now. Some design needs to be refactored later.

**Comment by sczyh30** [comment]
As this PR contains two parts of features, could you please separate it into two commits? One for Sentinel data property and one for enhancement of statistic module.

**Comment by sczyh30** [comment]
Thanks!

**Pull request merged**